docs: Add BUILD documentation for submitting app (#297)
WorldHealthOrganization/app: docs/BUILD.md

# Building the App for App Store Submission

> First, clone the repo (requires `git` - see [Dependencies](#dependencies)).  In order to ensure reproducibility, you will always build from a specified `TAG_NAME`.

Reviewer: Let's have them get all deps first.

```suggestion
First, make sure you have up-to-date versions of all [Dependencies](#dependencies) installed.

Second, clone the repo.  In order to ensure reproducibility, you will always build from a specified `TAG_NAME`.
```

```
git clone -b TAG_NAME https://github.com/WorldHealthOrganization/app.git
cd app
```

## iOS

Adapting instructions from [Flutter's iOS deployment docs](https://flutter.dev/docs/deployment/ios).

These instructions assume that an app record has already been added to the App Store Connect account. If you haven't done so already, you can follow [these instructions](https://help.apple.com/app-store-connect/#/dev2cd126805).

> ### Install Dependencies

Reviewer: ```suggestion
### Install CocoaPods Dependencies
```
to differentiate

> CocoaPods install (requires `CocoaPods` and `Flutter` - see [Dependencies](#dependencies)):

Reviewer: ```suggestion
CocoaPods install:
```

```
cd $(git rev-parse --show-toplevel)/client/flutter
flutter pub get
cd ios
pod install
```

### Review Xcode Build Settings

For detailed procedures and descriptions, see [Prepare for app distribution](https://help.apple.com/xcode/mac/current/#/dev91fe7130a).

Open the Xcode Workspace:

```

> open Runner.xcworkspace # run this in the client/flutter/ios directory

Reviewer: ```suggestion
cd $(git rev-parse --show-toplevel)/client/flutter/ios && open Runner.xcworkspace
```

```

In Xcode:

1. Select **Runner** in the Xcode project navigator, then select the **Runner** target in the settings view sidebar.
1. Select the General tab.
1. Do not change the **Display Name** from `Runner`
1. Verify you have chosen the same **Bundle Identifier** as the app you have registered on App Store Connect.
1. In the **Signing** section, ensure `Automatically manage signing` is selected, and select the team associated with your Apple Developer account. If other ways of signing code are needed, review the [Code Signing Guide](https://developer.apple.com/library/archive/documentation/Security/Conceptual/CodeSigningGuide/Introduction/Introduction.html) and [Create, export, and delete signing certificates](https://help.apple.com/xcode/mac/current/#/dev154b28f09)

### Create a Build Archive

Configure the app version and build:

1. Select **Product > Scheme > Runner**.
1. Select **Product > Destination > Generic iOS Device**.
1. Select **Runner** in the Xcode project navigator, then select the **Runner** target in the settings view sidebar.

> 1. In the Identity section, update the **Version** number to the user-facing version number you wish to publish.

Reviewer: These don't apply to flutter:
<img width="400" alt="Screen Shot 2020-03-25 at 3 32 41 AM" src="https://user-images.githubusercontent.com/11729521/77527423-4fe07700-6e49-11ea-9c03-d29813e1e823.png">

1. Follow the on-screen instructions to add APKs or app bundles, describe what's new in this release, and name your release. Use the APK generated in [Create a release](#create-a-release).
2. To save any changes you make to your release, select Save.
3. When you've finished preparing your release, select Review.

#### Review and roll out

Prerequisite: Before you can roll out your release, make sure you've completed your app's store listing, content rating, & pricing & distribution sections. When each section is complete, you'll see a green check mark next to it on the left menu.

Assets prepared for submission: https://drive.google.com/drive/folders/17wi6q3Vlpt9KB6FuEOpZBdCJHtLCSXzh?usp=sharing

1. Go to your Play Console.
2. Select an app.
3. On the left menu, select **Release management > App releases**.
4. Next to the release you want to roll out, select **Edit release**.
5. Review your draft release and make any additional changes that are needed.
6. Select **Review**. You'll be taken to the "Review and roll out release" screen, where you can make sure there aren't any issues with your release before rolling out to users.
7. Review any warnings or errors.

> 8. Select **Confirm rollout**.

Reviewer: Pretty sure this pushes straight to production.  

> Select Confirm rollout.
If you're rolling out your app's first release on production, clicking Confirm rollout will also publish your app to all Google Play users in the countries you selected.

Author: It's a lot harder imagining this in advance rather than actually doing it in the UI... but I'm pretty sure the last time I remember doing this, [this step](https://github.com/WorldHealthOrganization/app/pull/297/files#diff-05ce98cb03d150c05f7863b8b945a9aaR205) is what prevented the prod rollout?

Reviewer: I'm dumb :)

# Building the App for App Store Submission

> First, clone the repo (requires `git` - see [Dependencies](#dependencies)):

Reviewer: ```suggestion
First, clone the repo (requires `git` - see [Dependencies](#dependencies)).  In order to ensure reproducibility, you will always build from a specified `TAG_NAME`.
```

```

> git clone https://github.com/WorldHealthOrganization/app.git

Reviewer: ```suggestion
git clone -b TAG_NAME https://github.com/WorldHealthOrganization/app.git
```

Reviewer 2: Why use git? 
They can directly get the tag from here: 
`https://github.com/WorldHealthOrganization/app/archive/{{TAG}}.tar.gz`

cd app
```

## iOS

Adapting instructions from [Flutter's iOS deployment docs](https://flutter.dev/docs/deployment/ios).

These instructions assume that an app record has already been added to the App Store Connect account. If you haven't done so already, you can follow [these instructions](https://help.apple.com/app-store-connect/#/dev2cd126805).

### Install Dependencies

> `cd` into the `client/flutter/ios` directory and install CocoaPods dependencies (requires `CocoaPods` - see [Dependencies](#dependencies)):

Reviewer 3: `cd` is redundant as it's listed in the command below, just keep it high level:

```suggestion
CocoaPods install - see [Dependencies](#dependencies):
```

```
cd client/flutter/ios

> pod install

Reviewer 3: @advayDev1 - should we require a specific cocoapods install?

Reviewer: See comment at bottom

1. In the sidebar of the Xcode Organizer window, select your iOS app, then select the build archive you just produced.
1. Click the **Validate App** button. If any issues are reported, address them and produce another build. You can reuse the same build ID until you upload an archive.
1. After the archive has been successfully validated, click **Distribute App**. You can follow the status of your build in the Activities tab of your app’s details page on App Store Connect.

You should receive an email within 30 minutes notifying you that your build has been validated.

### Submit For Review

Go to the app's application details page on App Store Connect, and follow these steps:

1. Select Pricing and Availability from the sidebar of your app’s application details page on App Store Connect and complete the required information.
1. Select the status from the sidebar and complete all required fields. [Assets prepared for submission](https://drive.google.com/drive/folders/17wi6q3Vlpt9KB6FuEOpZBdCJHtLCSXzh?usp=sharing)
1. Click **Submit for Review**.

## Android

Adapting instructions from [Flutter's Android deployment docs](https://flutter.dev/docs/deployment/android)

> ### Signing the App

Reviewer 3: Replace lines 75 to 170 with just a link to the signing steps:
https://flutter.dev/docs/deployment/android#signing-the-app

   if (keystorePropertiesFile.exists()) {
       keystoreProperties.load(new FileInputStream(keystorePropertiesFile))
   }

   android {
         ...
   }
```

Load the `key.properties` file into the `keystoreProperties` object.

2. Add code before buildTypes block:

```
   buildTypes {
       release {
           // TODO: Add your own signing config for the release build.
           // Signing with the debug keys for now,
           // so `flutter run --release` works.
           signingConfig signingConfigs.debug
       }
   }
```

With the signing configuration info:

```
   signingConfigs {
       release {
           keyAlias keystoreProperties['keyAlias']
           keyPassword keystoreProperties['keyPassword']
           storeFile keystoreProperties['storeFile'] ? file(keystoreProperties['storeFile']) : null
           storePassword keystoreProperties['storePassword']
       }
   }
   buildTypes {
       release {
           signingConfig signingConfigs.release
       }
   }
```

Configure the `signingConfigs` block in your module’s `build.gradle` file.

Release builds of your app will now be signed automatically.

> ### Reviewing the app manifest

Reviewer 3: remove this section as the app manifest is already checked by the engineers

Review the default [App Manifest](https://developer.android.com/guide/topics/manifest/manifest-intro) file, `AndroidManifest.xml`, located in `client/flutter/android/app/src/main` and verify that the values are correct, especially the following:

`application`

> Edit the `android:label` in the application tag to reflect the final name of the app.

Reviewer: They should tell us the app name, we should set it and never touch it again.

Author: removed

> ### Reviewing the build configuration

Reviewer 3: remove this too, same reason as above

Review the default [Gradle build file](https://developer.android.com/studio/build/#module-level) file, `build.gradle`, located in `client/flutter/android/app` and verify the values are correct, especially the following values in the `defaultConfig` block:

`applicationId`

> Specify the final, unique (Application Id)appid

Reviewer: They should tell us the app id, we should set it and never touch it again.

Author: removed whole section

`versionCode` & `versionName`

> Specify the internal app version number, and the version number display string. You can do this by setting the version property in the pubspec.yaml file. Consult the version information guidance in the [versions documentation](https://developer.android.com/studio/publish/versioning).

Reviewer: They should not be setting the version numbers, we should be.

Author: removing this whole section if we will set and send to them

### Building the app for release

These instructions use the App Bundle release format instead of the APK format.

#### Build an app bundle

> `cd` into the `client/flutter` directory:

Reviewer 3: simplify description:
> Move to `client/flutter` and build app:
> ```
> #assumes top level in repository
> cd client/flutter
> flutter build appbundle
> ```

replaces line 198 to 208

> `cd` into the `client/flutter` directory:

Reviewer: ```suggestion
These steps must be performed from the `client/flutter` directory:
```

```

> cd .. # this assumes you're in the client/flutter/ios directory, if you're in the app directory you'll want to cd client/flutter

Reviewer 3: TMI, assume top level directory

> cd .. # this assumes you're in the client/flutter/ios directory, if you're in the app directory you'll want to cd client/flutter

Reviewer: ```suggestion
cd $(git rev-parse --show-toplevel)/client/flutter
```

```

> Build the app bundle with flutter:

Reviewer: delete this

```
flutter build appbundle
```

The release bundle for your app is created at `client/flutter/build/app/outputs/bundle/release/app.aab`.

### Upload the bundle to Google Play

> This adapts instructions found [here](https://support.google.com/googleplay/android-developer/answer/7159011).

Reviewer 3: Replace lines 214 to 247 with just the link to the Google Play instructions. This follows the important DRY principle: https://en.wikipedia.org/wiki/Don%27t_repeat_yourself

With a copy, the problem is when the instructions change. Google's documentation is automatically updated but the documentation here is not. So the goal is to have the WHO app specific comments here and then refer to the other documentation elsewhere

Reviewer: Most important difference to normal releases is:
> 4. Next to the release type you want to create, select **Manage**, then select **Closed**.

to keep it from going live

2. To save any changes you make to your release, select Save.
3. When you've finished preparing your release, select Review.

#### Review and roll out

Prerequisite: Before you can roll out your release, make sure you've completed your app's store listing, content rating, & pricing & distribution sections. When each section is complete, you'll see a green check mark next to it on the left menu.

Assets prepared for submission: https://drive.google.com/drive/folders/17wi6q3Vlpt9KB6FuEOpZBdCJHtLCSXzh?usp=sharing

1. Go to your Play Console.
2. Select an app.
3. On the left menu, select **Release management > App releases**.
4. Next to the release you want to roll out, select **Edit release**.
5. Review your draft release and make any additional changes that are needed.
6. Select **Review**. You'll be taken to the "Review and roll out release" screen, where you can make sure there aren't any issues with your release before rolling out to users.
7. Review any warnings or errors.
8. Select **Confirm rollout**.

## Dependencies

* [`git`](https://git-scm.com/book/en/v2/Getting-Started-Installing-Git)

> * [CocoaPods](https://guides.cocoapods.org/using/getting-started.html#installation) (for building iOS)

Reviewer: Require the latest cocoapods specifically and no earlier version (seeing all the issues both of us had with that)

# Building the App for App Store Submission

These instructions will assume that you have the up-to-date [`git`](https://git-scm.com/book/en/v2/Getting-Started-Installing-Git), [Flutter](https://flutter.dev/docs/get-started/install), and [CocoaPods](https://guides.cocoapods.org/using/getting-started.html#installation) already installed.
  * _Note: this was tested using CocoaPods version `1.9.1`_

## Clone the Repo

> In order to ensure reproducibility, you will always build from a specified `TAG_NAME`.

Reviewer: ```suggestion
In order to ensure reproducibility, you will always build from a specified `TAG_NAME` and should not change bundle/application ids or version or build numbers that are stored in the git repository.
```

Follow Apple's instructions to [create an archive](https://help.apple.com/xcode/mac/current/#/devf37a1db04), [validate the app](https://help.apple.com/xcode/mac/current/#/dev37441e273), and [upload it to App Store Connect](https://help.apple.com/xcode/mac/current/#/dev442d7f2ca).

You should receive an email within 30 minutes notifying you that your build has been validated.

### Submit For Review

Follow Apple's instructions to [publish the app](https://help.apple.com/app-store-connect/#/dev34e9bbb5a). We have prepared the [assets needed for submission](https://drive.google.com/drive/folders/17wi6q3Vlpt9KB6FuEOpZBdCJHtLCSXzh?usp=sharing).

## Android

### Signing the App

Follow Flutter's [instructions for signing the app](https://flutter.dev/docs/deployment/android#signing-the-app).

### Building the app for release

Build the app bundle:

```
cd $(git rev-parse --show-toplevel)/client/flutter
flutter build appbundle
```

The release bundle for your app is created at `client/flutter/build/app/outputs/bundle/release/app.aab`.

### Upload the bundle to Google Play

> Follow Google's instructions to [prepare & roll out releases](https://support.google.com/googleplay/android-developer/answer/7159011). We have prepared the [assets needed for submission](https://drive.google.com/drive/folders/17wi6q3Vlpt9KB6FuEOpZBdCJHtLCSXzh?usp=sharing).

Reviewer: ```suggestion
Follow Google's instructions to [prepare & roll out releases](https://support.google.com/googleplay/android-developer/answer/7159011).

**WARNING:** You must set the release type you want to create to "Closed" or you will inadvertently distribute the app to the general public.

We have prepared the [assets needed for submission](https://drive.google.com/drive/folders/17wi6q3Vlpt9KB6FuEOpZBdCJHtLCSXzh?usp=sharing).
```

# Building the App for App Store Submission

These instructions will assume that you have the up-to-date [`git`](https://git-scm.com/book/en/v2/Getting-Started-Installing-Git), [Flutter](https://flutter.dev/docs/get-started/install), and [CocoaPods](https://guides.cocoapods.org/using/getting-started.html#installation) already installed.
  * _Note: this was tested using CocoaPods version `1.9.1`_

## Clone the Repo

> In order to ensure reproducibility, you will always build from a specified `TAG_NAME` and should not change bundle/application ids or version or build numbers that are stored in the git repository.

Reviewer 3: Add example tag, this is fine as the commands still stay generic:
> `TAG_NAME` (e.g. v0.1.0)

flutter pub get
cd ios
pod install
```

### Prepare for app distribution

Follow Apple's instructions on [preparing for app distribution](https://help.apple.com/xcode/mac/current/#/dev91fe7130a) and [code signing](https://developer.apple.com/library/archive/documentation/Security/Conceptual/CodeSigningGuide/Introduction/Introduction.html).

**_Note: make sure you do not change the Display Name in the Runner target; Display Name should remain `Runner`_**

### Create a Build Archive

Prepare the Flutter release:

```
cd $(git rev-parse --show-toplevel)/client/flutter
flutter build ios
```

_Note: if you are on Xcode version 8.2 or earlier, you will need to close and re-open Xcode to ensure that the release configuration mode is refreshed_

Follow Apple's instructions to [create an archive](https://help.apple.com/xcode/mac/current/#/devf37a1db04), [validate the app](https://help.apple.com/xcode/mac/current/#/dev37441e273), and [upload it to App Store Connect](https://help.apple.com/xcode/mac/current/#/dev442d7f2ca).

You should receive an email within 30 minutes notifying you that your build has been validated.

### Submit For Review

> Follow Apple's instructions to [publish the app](https://help.apple.com/app-store-connect/#/dev34e9bbb5a). We have prepared the [assets needed for submission](https://drive.google.com/drive/folders/17wi6q3Vlpt9KB6FuEOpZBdCJHtLCSXzh?usp=sharing).

Reviewer: **WARNING:** Choose the "Release the app yourself" or "Manually release this version" option in the "Version Release" section

### Create a Build Archive

Prepare the Flutter release:

```
cd $(git rev-parse --show-toplevel)/client/flutter
flutter build ios
```

_Note: if you are on Xcode version 8.2 or earlier, you will need to close and re-open Xcode to ensure that the release configuration mode is refreshed_

Follow Apple's instructions to [create an archive](https://help.apple.com/xcode/mac/current/#/devf37a1db04), [validate the app](https://help.apple.com/xcode/mac/current/#/dev37441e273), and [upload it to App Store Connect](https://help.apple.com/xcode/mac/current/#/dev442d7f2ca).

You should receive an email within 30 minutes notifying you that your build has been validated.

### Submit For Review

Follow Apple's instructions to [publish the app](https://help.apple.com/app-store-connect/#/dev34e9bbb5a). We have prepared the [assets needed for submission](https://drive.google.com/drive/folders/17wi6q3Vlpt9KB6FuEOpZBdCJHtLCSXzh?usp=sharing).

**WARNING:** In step 4, choose the  "Manually release this version" option in the "Version Release" section to prevent inadvertent public distribution.

## Android

### Signing the App

> Follow Flutter's [instructions for signing the app](https://flutter.dev/docs/deployment/android#signing-the-app).

Reviewer: ```suggestion
Follow Flutter's [instructions for signing the app](https://flutter.dev/docs/deployment/android#signing-the-app).

Consider using [App Signing by Google Play](https://github.com/WorldHealthOrganization/app/issues/300) to manage your signing keys.
```

# Building the App for App Store Submission

First, clone the repo (requires `git` - see [Dependencies](#dependencies)):

```
git clone https://github.com/WorldHealthOrganization/app.git
cd app
```

## iOS

These instructions assume that an app record has already been added to the App Store Connect account. If you haven't done so already, you can follow [these instructions](https://help.apple.com/app-store-connect/#/dev2cd126805).

### Install Dependencies

`cd` into the `client/flutter/ios` directory:

```
cd client/flutter/ios
```

Install CocoaPods dependencies (requires `CocoaPods` - see [Dependencies](#dependencies)):

```

> pod install

Reviewer: ````
$ cd client/flutter/ios
$ pod install
Ignoring iStats-1.6.1 because its extensions are not built.  Try: gem pristine iStats --version 1.6.1

[!] Invalid `Podfile` file: Generated.xcconfig must exist. If you're running pod install manually, make sure flutter pub get is executed first.
````

Author: what is your version of ruby and CocoaPods? it looks like they'll have to install flutter for Android anyway, so this may not be a problem, but just checking

Reviewer: $ ruby --version
ruby 2.3.7p456 (2018-03-28 revision 63024) [universal.x86_64-darwin18]
$ pod --version
Ignoring iStats-1.6.1 because its extensions are not built.  Try: gem pristine iStats --version 1.6.1
1.9.1

Author: also, Pat prepared #289 to check the dependencies directly into the repo just in case there were any challenges with CocoaPods. opens up the question of storing the dependencies here, which I'm not sure how you feel about. but it would definitely make it easiest for them

Author: hmm... worked fine for me without this step with ruby `2.7.0` and CocoaPods `1.9.1`

Reviewer: works with `flutter pub get` in client/flutter

> ```

Reviewer 3: Make this more concise:

> ### Install Dependencies
> Install CocoaPods dependencies (requires `CocoaPods` - see [Dependencies] (#dependencies)):
> ```
> cd client/flutter/ios
> pod install
> ```

> Open the Xcode Workspace:

Reviewer 3: Move this in to next section in Xcode Build Settings

For detailed procedures and descriptions, see [Prepare for app distribution](https://help.apple.com/xcode/mac/current/#/dev91fe7130a).

In Xcode:

1. Select **Runner** in the Xcode project navigator, then select the **Runner** target in the settings view sidebar.
1. Select the General tab.
1. Do not change the **Display Name** from `Runner`
1. Verify you have chosen the same **Bundle Identifier** as the app you have registered on App Store Connect.
1. In the **Signing** section, ensure `Automatically manage signing` is selected, and select the team associated with your Apple Developer account. If other ways of signing code are needed, review the [Code Signing Guide](https://developer.apple.com/library/archive/documentation/Security/Conceptual/CodeSigningGuide/Introduction/Introduction.html) and [Create, export, and delete signing certificates](https://help.apple.com/xcode/mac/current/#/dev154b28f09)

### Create a Build Archive

Configure the app version and build:

1. Select **Product > Scheme > Runner**.
1. Select **Product > Destination > Generic iOS Device**.
1. Select **Runner** in the Xcode project navigator, then select the **Runner** target in the settings view sidebar.

> 1. In the Identity section, update the **Version** number to the user-facing version number you wish to publish. For the initial App Store submission, version number should be `0.1.0`.

Reviewer 3: remove reference to `0.1.0`, that version number should be shared over email and not embedded in instructions where it becomes stale

1. In the Identity section, update the **Build** identifier to a unique build number used to track this build on App Store Connect. Each upload requires a unique build number.

Create a build archive and upload it to App Store Connect:

1. Select **Product > Archive** to produce a build archive.
1. In the sidebar of the Xcode Organizer window, select your iOS app, then select the build archive you just produced.
1. Click the **Validate App** button. If any issues are reported, address them and produce another build. You can reuse the same build ID until you upload an archive.
1. After the archive has been successfully validated, click **Distribute App**. You can follow the status of your build in the Activities tab of your app’s details page on App Store Connect.

You should receive an email within 30 minutes notifying you that your build has been validated.

### Submit For Review

Go to the app's application details page on App Store Connect, and follow these steps:

1. Select Pricing and Availability from the sidebar of your app’s application details page on App Store Connect and complete the required information.

>   * **_TODO: fill in details for how to not publish_**

Author: I know formatting is off in this section, I will fix when adding Android

> 1. Select the status from the sidebar. If this is the first release of this app, its status is 1.0 Prepare for Submission. Complete all required fields.

Reviewer 3: "its status is 1.0 Prepare for Submission" didn't make sense to me as a sentence, not sure what the intent was

  * Assets prepared for submission: https://drive.google.com/drive/folders/17wi6q3Vlpt9KB6FuEOpZBdCJHtLCSXzh?usp=sharing
1. Click **Submit for Review**.

## Android

> 

Reviewer 3: "TODO" is fine for now
